pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+		return
+	}
+	newQuota := toElasticQuota(newObj)
+	if newQuota == nil {
+		klog.Errorf("cannot convert to *v1.Quota:%v", newObj)
+		return
+	}
+
+	oldQuotaInfo := core.NewQuotaInfoFromQuota(oldQuota)
+	newQuotaInfo := core.NewQuotaInfoFromQuota(newQuota)
+
+	qt.quotaInfoMap[newQuotaInfo.Name] = newQuotaInfo
+	// parentQuotaName change
+	if oldQuotaInfo.ParentName != newQuotaInfo.ParentName {
+		delete(qt.quotaHierarchyInfo[oldQuotaInfo.ParentName], oldQuotaInfo.Name)
+		if qt.quotaHierarchyInfo[newQuotaInfo.ParentName] == nil {
+			qt.quotaHierarchyInfo[newQuotaInfo.ParentName] = make(map[string]struct{})
+		}
+		qt.quotaHierarchyInfo[newQuotaInfo.ParentName][newQuotaInfo.Name] = struct{}{}
+	}
+}
+
+func (qt *quotaTopology) OnQuotaDelete(obj interface{}) {
+	quota := toElasticQuota(obj)
+	if quota == nil {
+		klog.Errorf("cannot convert to *v1.Quota:%v", obj)
+		return
+	}
+
+	parentName := extension.GetParentQuotaName(quota)
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	if parentName != extension.RootQuotaName {
+		delete(qt.quotaHierarchyInfo[parentName], quota.Name)
+	}
+	delete(qt.quotaHierarchyInfo, quota.Name)
+	delete(qt.quotaInfoMap, quota.Name)
+	delete(qt.podCache, quota.Name)
+}
+
+func (qt *quotaTopology) OnPodAdd(obj interface{}) {
+	pod := obj.(*v1.Pod)
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	quotaName := qt.getQuotaNameFromPodNoLock(pod)
+	qt.podCache[quotaName][pod.Name] = struct{}{}
+	klog.V(5).Infof("onPodAdd, link pod :%v to quota:%v", pod.Name, quotaName)
+}
+
+func (qt *quotaTopology) OnPodUpdate(oldObj, newObj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	newPod := newObj.(*v1.Pod)
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	oldQuotaName := qt.getQuotaNameFromPodNoLock(oldPod)
+	newQuotaName := qt.getQuotaNameFromPodNoLock(newPod)
+	delete(qt.podCache[oldQuotaName], oldPod.Name)
+	qt.podCache[newQuotaName][newPod.Name] = struct{}{}
+}
+
+func (qt *quotaTopology) OnPodDelete(obj interface{}) {
+	pod := obj.(*v1.Pod)
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	quotaName := qt.getQuotaNameFromPodNoLock(pod)
+	delete(qt.podCache[quotaName], pod.Name)
+}
+
+func (qt *quotaTopology) AddQuota(quota *v1alpha1.ElasticQuota) (err error) {
+	if quota == nil {
+		return fmt.Errorf("AddQuota param is nil")
+	}
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	if _, exist := qt.quotaInfoMap[quota.Name]; exist {
+		return fmt.Errorf("quota already exist:%v", quota.Name)
+	}
+
+	// fill quota with basic Information
+	err = qt.fillDefaultQuotaInfo(quota)
+	if err != nil {
+		klog.Errorf("AddQuota fillDefaultQuotaInfo failed:%v", err)
+		return err
+	}
+
+	err = qt.basicItemCheck(quota)
+	if err != nil {
+		klog.Errorf("AddQuota basicItemCheck failed :%v", err)
+		return err
+	}
+
+	quotaInfo := core.NewQuotaInfoFromQuota(quota)
+
+	err = qt.validateQuota(quotaInfo)
+	if err != nil {
+		klog.Errorf("AddQuota validateQuota failed:%v", err)
+		return err
+	}
+
+	qt.quotaInfoMap[quotaInfo.Name] = quotaInfo
```

Comment:
what's the relationship with these code with OnQuotaAdd(), it seems the job is same.  I prefer package same codes to a function